### PR TITLE
Introduce ruff linter

### PR DESCRIPTION
**Pull request by phschiele**
## Description
This PR adds configuration for [ruff](https://github.com/charliermarsh/ruff), an extremely fast linter for Python projects and is used by SciPy, pandas, and many others. It includes the relevant rules from flake8 and isort.
Currently, the `isort` pre-commit hook needs an upgrade and I used this as an opportunity to unify all of our linting under one umbrella.
You can run
```
pip install ruff
ruff check cvxpy
```
to try it out (yes, it actually did execute). 

To update you're pre-commit hooks in this branch, run
```
pre-commit clean
pre-commit install
```
Mostly put this together out of curiosity and so other people can try it out easily to see if they like it. 

Seems to work nicely with the GitHub inline errors:
image

Tasks:
- [x] Configuration
- [x] Documentation
- [x] Linter action

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [x] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [ ] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [ ] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

**Comment by phschiele** [comment]
Updated the docs:
image

File: pyproject.toml [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+line-length = 100
+exclude = [
+    "build",
+    "examples",
+    "doc",
+    "cvxpy/cvxcore/*",
+    "*__init__.py"
 ]
+# Assume Python 3.8.
+target-version = "py38"
```
Comment (SteveDiamond) [inline pyproject.toml:12]: What is this about?

Comment (phschiele) [reply pyproject.toml:12]: It is the "minimum Python version that should be supported". Was copied from another code base, changed it to 3.7.

Comment (phschiele) [inline pyproject.toml:12]:
```suggestion
# The minimum Python version that should be supported
target-version = "py37"
```

Comment (phschiele) [reply pyproject.toml]: I think this is to keep track of which Python features should be supported (like f-strings only with Python 3.6 etc).

**Pull request merged**